### PR TITLE
Serialization and representation GP + Small fixes

### PR DESCRIPTION
**Pull request by stevetorr**
Introduces a serialization method for GPs (to_ and from_dict methods) and a representation (__str__) method. Introduced a new string-to-kernel method in kernels.py to facilitate the from_dict method. The serialization is unit tested. Fixed a few style errors in other parts of the code. A method in gp.py sets a 'self.like' attribute and others set `self.likelihood`. This is also now corrected towards `self.likelihood`.

### CODE REVIEW

**Comment by stevetorr** [comment]
Closes #63 , addresses #16

**Comment by dmclark17** [comment]
Regarding the string to kernel logic, you may be able to refactor the function to use `getattr` and `getattr`. You might be able to get away with not needing to define a dictionary mapping strings to functions. 

Unclear if this would be better though 🤷‍♂

**Comment by stevetorr** [comment]
> `numpy.ndarray`s are not JSON serializable using the default encoder. Is there a canonical way to store them on disk for the project?

I've been using an encoder that I found online which is included in util.py to handle converting numpy arrays to JSON.


> Regarding the string to kernel logic, you may be able to refactor the function to use `getattr` and `getattr`. You might be able to get away with not needing to define a dictionary mapping strings to functions.

Thanks for looking closely at the methods! What would the `getattr` correspond to- getattr called on the file itself (to refer to individual callables within the flare.kernels file)? Also can you clarify what the second `getattr` was supposed to be?

**Comment by dmclark17** [comment]
I see, I didn't see the `NumpyEncoder` 👍 


Whoops that was suppose to be `getattr` and `hasattr`. You can call them on the module itself by doing something like:

```python
current_module = sys.modules[__name__]  # Gets the current module
if hasattr(current_module, "two_body"):
    func = getattr(current_module, "two_body")
```

File: flare/gp.py [hunk 1]
```diff
@@ -185,12 +195,6 @@ def predict(self, x_t: AtomicEnvironment, d: int) -> [float, float]:
         pred_var = self_kern - \
                    np.matmul(np.matmul(k_v, self.ky_mat_inv), k_v)
 
-        # # get predictive variance (possibly slow)
-        # v_vec = solve_triangular(self.l_mat, k_v, lower=True)
-        # self_kern = self.kernel(x_t, x_t, self.bodies, d, d, self.hyps,
-        #                         self.cutoffs)
-        # pred_var = self_kern - np.matmul(v_vec, v_vec)
-
```
Comment (jonpvandermause) [inline flare/gp.py]: Nice, these lines needed to go.

File: flare/gp.py [hunk 2]
```diff
@@ -292,8 +296,8 @@ def set_L_alpha(self):
         self.ky_mat_inv = ky_mat_inv
         self.l_mat_inv = l_mat_inv
 
-        self.like = like
-        self.like_grad = like_grad
+        self.likelihood = like
+        self.likelihood_gradient = like_grad
```
Comment (jonpvandermause) [inline flare/gp.py]: Good catch, not sure why there were multiple versions of the likelihood.

File: flare/gp.py [hunk 3]
```diff
@@ -331,3 +335,82 @@ def update_L_alpha(self):
...
+        thestr += 'Hyperparameters: \n'
+        if self.hyp_labels is None:
+            # Put unlabeled hyperparameters on one line
+            thestr = thestr[:-1]
+            thestr += str(self.hyps) + '\n'
+        else:
+            for hyp, label in zip(self.hyps, self.hyp_labels):
+                thestr += "{}: {}\n".format(label, hyp)
+
+        return thestr
```
Comment (jonpvandermause) [inline flare/gp.py]: Slick.

**Comment by stevetorr** [comment]
Thanks for taking the time to look through and provide feedback!

**Pull request merged**